core-backend/src/funcs.rs

```diff
@@ -24,67 +24,309 @@ use crate::{
-            s.push_str(&syscall_args_trace!($($rest),+));
-            s
+    }
+}
+
+struct ArgsFormatter<'a>(&'a [Value]);
+
+impl fmt::Display for ArgsFormatter<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut iter = self.0.iter();
+
+        if let Some(value) = iter.next() {
+            write!(f, "{}", ValueFormatter(value))?;
         }
-    };
+
+        for value in iter {
+            write!(f, ", {}", ValueFormatter(value))?;
+        }
+
+        Ok(())
+    }
 }
 
-macro_rules! syscall_trace {
-    ($name:expr, $($args:expr),+) => {
-        {
-            ::log::trace!(target: "syscalls", "{}{}", $name, syscall_args_trace!($($args),+));
+fn function_name<T>() -> &'static str {
+    let s = core::any::type_name::<T>();
+    let pos = s.rfind("::").unwrap();
+    &s[pos + 2..]
+}
+
+/// Actually just wrapper around [`Value`] to implement conversions.
+#[derive(Clone, Copy)]
+struct SysCallValue(Value);
+
+impl From<i32> for SysCallValue {
+    fn from(value: i32) -> Self {
+        SysCallValue(Value::I32(value))
+    }
+}
+
+impl From<u32> for SysCallValue {
+    fn from(value: u32) -> Self {
+        SysCallValue(Value::I32(value as i32))
+    }
+}
+
+impl From<i64> for SysCallValue {
+    fn from(value: i64) -> Self {
+        SysCallValue(Value::I64(value))
+    }
+}
+
+impl TryFrom<SysCallValue> for u32 {
+    type Error = HostError;
+
+    fn try_from(val: SysCallValue) -> Result<u32, HostError> {
+        if let Value::I32(val) = val.0 {
+            Ok(val as u32)
+        } else {
+            Err(HostError)
         }
-    };
-    ($name:expr) => {
+    }
+}
+
+impl TryFrom<SysCallValue> for u64 {
+    type Error = HostError;
+
+    fn try_from(val: SysCallValue) -> Result<u64, HostError> {
+        if let Value::I64(val) = val.0 {
+            Ok(val as u64)
+        } else {
+            Err(HostError)
+        }
+    }
+}
+
+/// Actually just wrapper around [`ReturnValue`] to implement conversions.
+pub struct SysCallReturnValue(ReturnValue);
+
+impl From<SysCallReturnValue> for ReturnValue {
+    fn from(value: SysCallReturnValue) -> Self {
+        value.0
+    }
+}
+
+impl From<()> for SysCallReturnValue {
+    fn from((): ()) -> Self {
+        Self(ReturnValue::Unit)
+    }
+}
+
+impl From<i32> for SysCallReturnValue {
+    fn from(value: i32) -> Self {
+        Self(ReturnValue::Value(Value::I32(value)))
+    }
+}
+
+impl From<u32> for SysCallReturnValue {
+    fn from(value: u32) -> Self {
+        Self(ReturnValue::Value(Value::I32(value as i32)))
+    }
+}
+
+pub(crate) trait SysCallContext: Sized {
+    fn from_args(args: &[Value]) -> Result<(Self, &[Value]), HostError>;
+}
+
+pub(crate) trait SysCall<Ext, T = ()> {
+    type Context: SysCallContext;
+
+    fn execute(
+        self,
+        caller: &mut CallerWrap<Ext>,
+        ctx: Self::Context,
+    ) -> Result<(u64, T), HostError>;
+}
+
+pub(crate) trait SysCallBuilder<Ext, Args: ?Sized, R, S> {
```

Comment:
I'd remove these unused and unintuitive generics maybe in favour of associated types if needed? And expand naming for them please